src/id_allocator.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! Provides allocation and releasing strategy for IDs.
+//!
+//! This module implements an allocation strategies for all resource types
+//! that can be abstracted to an integer.
+
+use std::collections::BTreeSet;
+use std::fmt;
+use std::result;
+
+/// Error type for IdAllocator usage.
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// All ids from the range specified are exhausted.
+    Overflow,
+    /// An id that is not part of the specified range was requested to be released.
+    InvalidInput(u32),
+    /// There are no more IDs available in the manage range
+    ResourceExhausted,
+}
+
+pub type Result = result::Result<u32, Error>;
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match *self {
+            Overflow => write!(f, "Id counter overflowed."),
+            InvalidInput(id) => write!(f, "Specified id: {} is not in the range.", id),
+            ResourceExhausted => {
+                write!(f, "There are no more available ids in the specified range.")
+            }
+        }
+    }
+}
+
+/// # Example
+///
+/// ```rust
+/// # use vm_allocator::IdAllocator;
+///
+/// let mut legacy_irq_allocator = IdAllocator::new(5, 23);
+///
+/// let irq_nr = legacy_irq_allocator.allocate_id();
+/// if irq_nr.is_err() {
+///     println!("Interrupt number allocation failed: {:?}", irq_nr);
+/// }
+/// let irq = 8;
+/// let freed_id_result = legacy_irq_allocator.free_id(irq);
+/// if freed_id_result.is_err() {
+///     println!("Failed to free the specified id: {:?}", freed_id_result);
+/// }
+/// ```
+
+/// Internal representation of IdAllocator. Contains the ends of the interval
+/// that is managed, a field that points at the next available ID, and a binary
+/// tree set used to store the released IDs. The reason we have chosed a
+/// BTreeSet is that the overage complexity for deletion and insertion is
+/// O(logN) compared to Vec for example, anothe benefit is that the entries
+/// are sorted so we will always use the first available ID.
+pub struct IdAllocator {
+    // Beginnbing of the range of IDs that we want to manage.
+    range_base: u32,
+    // First available id that was never allocated.
+    next_id: u32,
+    // End of the range of IDs that we want to manage.
+    range_end: u32,
+    // Set of all freed ids that can be reused at subsequent allocations.
+    freed_ids: BTreeSet<u32>,
+}
+
+impl IdAllocator {
+    /// Creates a new instance of IdAllocator that will be used to manage the
+    /// allocation and release of ids from the interval specified by
+    /// `range_base` and `range_max`
+    pub fn new(range_base: u32, range_max: u32) -> Self {
```

Comment:
We should use consistent naming for the range end struct field and parameter; now it's `range_end` in the struct and `range_max` in the constructor.